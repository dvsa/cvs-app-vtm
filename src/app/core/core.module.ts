import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { GlobalErrorComponent } from './components/global-error/global-error.component';
import { SpinnerComponent } from './components/spinner/spinner.component';
import { FooterComponent } from './components/footer/footer.component';
import { HeaderComponent } from './components/header/header.component';
import { RouterModule } from '@angular/router';
import { BreadcrumbsComponent } from './components/breadcrumbs/breadcrumbs.component';
import { PageNotFoundComponent } from './components/page-not-found/page-not-found.component';
import { ServerErrorComponent } from './components/server-error/server-error.component';

@NgModule({
  declarations: [FooterComponent, HeaderComponent, GlobalErrorComponent, SpinnerComponent, BreadcrumbsComponent, PageNotFoundComponent, ServerErrorComponent],
  imports: [CommonModule, RouterModule],
  exports: [FooterComponent, HeaderComponent, GlobalErrorComponent, SpinnerComponent, BreadcrumbsComponent]
})
export class CoreModule {}
