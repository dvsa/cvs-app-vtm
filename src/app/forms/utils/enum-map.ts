import { MultiOptions } from '@forms/models/options.model';

export function getOptionsFromEnum(object: object): MultiOptions {
  return Object.values(object).map(value => ({ value, label: value.charAt(0).toUpperCase() + value.slice(1) }));
}

export function getOptionsFromEnumOneChar(object: object): MultiOptions {
  return Object.values(object).map(value => ({ value, label: value.charAt(0).toUpperCase() }));
}

export function getOptionsFromEnumAcronym(object: object): MultiOptions {
  return Object.values(object).map(value => ({ value, label: value.toUpperCase() }));
}
