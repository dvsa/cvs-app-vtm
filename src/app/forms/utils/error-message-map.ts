import { formatDate } from '@angular/common';
import { AsyncValidatorNames } from '@forms/models/async-validators.enum';
import { ValidatorNames } from '@forms/models/validators.enum';

const DEFAULT_LABEL = 'This field';
export const ErrorMessageMap: Record<string, (...args: any) => string> = {
  // Date errors
  invalidDate: (err: { error: boolean; reason: string; index: number }) => `${err.reason}`,
  invalidOption: (err: boolean, label?: string) => `${label || DEFAULT_LABEL} is invalid`,
  invalidTestResult: (err: { message: string }) => err.message,

  [ValidatorNames.AheadOfDate]: (err: { sibling: string; date: Date }, label?: string) =>
    `${label || 'This date'} must be ahead of ${err.sibling || 'the previous date'}${err.date ? formatDate(err.date, ' (dd/MM/yyyy)', 'en') : ''}`,
  [ValidatorNames.CustomPattern]: (err: { message: string }, label?: string) => `${label || DEFAULT_LABEL} ${err.message}`,
  [ValidatorNames.DateNotExceed]: (err: { sibling: string; months: number }, label?: string) =>
    `${label || 'This date'} must be less than ${err.months || 'X'} months after ${err.sibling || 'the previous date'}`,
  [ValidatorNames.Defined]: (err: boolean, label?: string) => `${label} is required`,
  [ValidatorNames.FutureDate]: (err: boolean, label?: string) => `${label || 'This date'} must be in the future`,
  [ValidatorNames.Max]: (err: { max: number }, label?: string) => `${label || DEFAULT_LABEL} must be less than or equal to ${err.max}`,
  [ValidatorNames.MaxLength]: (err: { requiredLength: number }, label?: string) =>
    `${label || DEFAULT_LABEL} must be less than or equal to ${err.requiredLength} characters`,
  [ValidatorNames.Min]: (err: { min: number }, label?: string) => `${label || DEFAULT_LABEL} must be greater than or equal to ${err.min}`,
  [ValidatorNames.MinLength]: (err: { requiredLength: number }, label?: string) =>
    `${label || DEFAULT_LABEL} must be greater than or equal to ${err.requiredLength} characters`,
  [ValidatorNames.NotZNumber]: () => "The VRM/Trailer ID cannot be in a format that is 7 digits followed by the character 'Z'",
  [ValidatorNames.PastDate]: (err: boolean, label?: string) => `${label || 'This date'} must be in the past`,
  [ValidatorNames.Pattern]: (err: boolean, label?: string) => `${label || DEFAULT_LABEL} must match a pattern`,
  [ValidatorNames.Required]: (err: boolean, label?: string) => `${label || DEFAULT_LABEL} is required`,
  [ValidatorNames.RequiredIfEquals]: (err: { sibling: string }, label?: string) => `${label || DEFAULT_LABEL} is required with ${err.sibling}`,
  [ValidatorNames.ValidateDefectNotes]: () => 'Notes is required',
  [ValidatorNames.ValidateProhibitionIssued]: () => 'Prohibition notice has not been issued.',

  [AsyncValidatorNames.RequiredIfNotAbandoned]: (err: boolean, label?: string) => `${label || DEFAULT_LABEL} is required`,
  [AsyncValidatorNames.RequiredIfNotFail]: (err: boolean, label?: string) => `${label || DEFAULT_LABEL} is required`,
  [AsyncValidatorNames.RequiredIfNotResult]: (err: boolean, label?: string) => `${label || DEFAULT_LABEL} is required`,
  [AsyncValidatorNames.RequiredIfNotResultAndSiblingEquals]: (err: boolean, label?: string) => `${label || DEFAULT_LABEL} is required`
};
