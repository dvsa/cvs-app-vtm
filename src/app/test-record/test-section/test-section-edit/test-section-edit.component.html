<dl class="govuk-summary-list">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">
      Test type
    </dt>
    <dd class="govuk-summary-list__value">
      {{ testType.testTypeName | DefaultNullOrEmpty | titlecase }}
    </dd>
    <dd class="govuk-summary-list__value">
      <a class="govuk-link" (click)="onChangeTestType()">
        <span>Change</span>
      </a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">
      Test code
    </dt>
    <dd class="govuk-summary-list__value">
      {{ testType.testCode | DefaultNullOrEmpty }}
    </dd>
    <dd class="govuk-summary-list__value"></dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">
      Test number
    </dt>
    <dd class="govuk-summary-list__value">
      {{ testType.testNumber | DefaultNullOrEmpty }}
    </dd>
    <dd class="govuk-summary-list__value"></dd>
  </div>
</dl>

<div formGroupName="testType">
  <div
    class="govuk-grid-row govuk-!-margin-bottom-3"
    [class.govuk-form-group--error]="
      isSubmitted && testResultChildForm.form.get('testResult')?.invalid
    "
  >
    <div class="govuk-grid-column-full">
      <label class="govuk-label govuk-!-font-weight-bold">
        Result
      </label>
      <span
        *ngIf="isSubmitted && testResultChildForm.form.get('testResult')?.invalid"
        class="govuk-error-message"
      >
        <span class="govuk-visually-hidden">Error:</span>
        <span>Select the test result</span>
      </span>
      <div class="govuk-radios">
        <div class="govuk-radios__item" *ngFor="let res of resultOptions">
          <input
            class="govuk-radios__input"
            id="{{ 'test-testResult-' + res }}"
            type="radio"
            formControlName="testResult"
            value="{{ res | lowercase }}"
            [checked]="testType.testResult !== 'prs' ? (testType.testResult | titlecase) == res : (testType.testResult | uppercase) == res"
            [class.govuk-input--error]="
              isSubmitted && testResultChildForm.form.get('testResult')?.invalid
            "
          />
          <label class="govuk-label govuk-radios__label">
            {{ res }}
          </label>
        </div>
      </div>
    </div>
  </div>

  <div *ngIf="isAbandoned" class="govuk-grid-row govuk-!-margin-bottom-5">
    <div class="govuk-grid-column-full">
      <div class="govuk-inset-text govuk-!-margin-bottom-0 govuk-!-padding-bottom-0">
        <div class="govuk-grid-row">
          <div class="govuk-grid-column-full">
            <h4 class="govuk-heading-s">Reason for abandoning</h4>

            <div class="govuk-checkboxes">
              <div
                *ngFor="let reason of reasonsForAbandoning; let i = index"
                class="govuk-checkboxes__item"
              >
                <input
                  class="govuk-checkboxes__input"
                  id="test-reason-{{ i }}"
                  formControlName="reasonForAbandoning"
                  type="checkbox"
                  value="{{ reason }}"
                />
                <label class="govuk-label govuk-checkboxes__label">
                  <span>{{ reason }}</span>
                </label>
              </div>
            </div>
          </div>
        </div>

        <div class="govuk-grid-row govuk-!-margin-top-7">
          <div class="govuk-grid-column-two-thirds">
            <div class="govuk-form-group">
              <label class="govuk-label govuk-!-font-weight-bold" for="additionalCommentsForAbandon">
                Additional details for abandoning
              </label>
              <span id="additionalCommentsForAbandon" class="govuk-hint">
                Notes about the reason for abandoning should be displayed here...
              </span>
              <textarea
                class="govuk-textarea"
                id="more-detail"
                formControlName="additionalCommentsForAbandon"
                rows="5"
                aria-describedby="additionalCommentsForAbandon"
                >{{ testType.additionalCommentsForAbandon }}</textarea
              >
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>

  <div
    *ngIf="applicableTestTypeIds1.hasOwnProperty(testType.testTypeId)"
    class="govuk-grid-row govuk-!-margin-bottom-7"
    [class.govuk-form-group--error]="
      isSubmitted && testResultChildForm.form.get('certificateNumber')?.invalid
    "
  >
    <div class="govuk-grid-column-one-half">
      <label class="govuk-label govuk-!-font-weight-bold">
        Certificate number
      </label>
      <span
        *ngIf="isSubmitted && testResultChildForm.form.get('certificateNumber')?.invalid"
        class="govuk-error-message"
      >
        <span class="govuk-visually-hidden">Error:</span>
        <span>Enter the certificate number</span>
      </span>
      <input
        class="govuk-input"
        id="test-certificateNumber"
        formControlName="certificateNumber"
        type="text"
        [class.govuk-input--error]="
          isSubmitted && testResultChildForm.form.get('certificateNumber')?.invalid
        "
      />
    </div>
  </div>

  <div
    *ngIf="applicableTestTypeIds1.hasOwnProperty(testType.testTypeId)"
    class="govuk-grid-row govuk-!-margin-bottom-3"
    [class.govuk-form-group--error]="
      isSubmitted && testResultChildForm.form.get('testExpiryDate')?.invalid
    "
  >
    <div class="govuk-grid-column-three-quarters">
      <label class="govuk-label govuk-!-font-weight-bold">
        Expiry Date
      </label>
      <span
        *ngIf="isSubmitted && testResultChildForm.form.get('testExpiryDate')?.invalid"
        class="govuk-error-message"
      >
        <span class="govuk-visually-hidden">Error:</span>
        <span>Enter the expiry date</span>
      </span>
      <vtm-date-input
        id="test-testExpiryDate"
        ariaDescribedBy="testExpiryDate"
        formControlName="testExpiryDate"
      >
      </vtm-date-input>
    </div>
  </div>

  <div
    *ngIf="applicableTestTypeIds1.hasOwnProperty(testType.testTypeId)"
    class="govuk-grid-row govuk-!-margin-bottom-3"
    [class.govuk-form-group--error]="
      isSubmitted && testResultChildForm.form.get('testAnniversaryDate')?.invalid
    "
  >
    <div class="govuk-grid-column-three-quarters">
      <label class="govuk-label govuk-!-font-weight-bold">
        Anniversary Date
      </label>
      <span
        *ngIf="isSubmitted && testResultChildForm.form.get('testAnniversaryDate')?.invalid"
        class="govuk-error-message"
      >
        <span class="govuk-visually-hidden">Error:</span>
        <span>Enter the anniversary date</span>
      </span>
      <vtm-date-input
        id="test-testAnniversaryDate"
        ariaDescribedBy="testAnniversaryDate"
        formControlName="testAnniversaryDate"
      >
      </vtm-date-input>
    </div>
  </div>

  <div
    class="govuk-grid-row govuk-!-margin-bottom-3"
    [class.govuk-form-group--error]="
      isSubmitted && testResultChildForm.form.get('testEndTimestamp')?.invalid
    "
  >
    <div class="govuk-grid-column-three-quarters">
      <label class="govuk-label govuk-!-font-weight-bold">
        Test date
      </label>
      <span
        *ngIf="isSubmitted && testResultChildForm.form.get('testEndTimestamp')?.invalid"
        class="govuk-error-message"
      >
        <span class="govuk-visually-hidden">Error:</span>
        <span>Enter the test date</span>
      </span>
      <vtm-date-input
        id="test-testEndTimestamp"
        ariaDescribedBy="testEndTimestamp"
        formControlName="testEndTimestamp"
      >
      </vtm-date-input>
    </div>
  </div>

  <div
    *ngIf="applicableTestTypeIds2.hasOwnProperty(testType.testTypeId)"
    class="govuk-grid-row govuk-!-margin-bottom-3"
    [class.govuk-form-group--error]="
      isSubmitted && testResultChildForm.form.get('prohibitionIssued')?.invalid
    "
  >
    <div class="govuk-grid-column-full">
      <label class="govuk-label govuk-!-font-weight-bold">
        Prohibition Issued
      </label>
      <span
        *ngIf="isSubmitted && testResultChildForm.form.get('prohibitionIssued')?.invalid"
        class="govuk-error-message"
      >
        <span class="govuk-visually-hidden">Error:</span>
        <span>Select the prohibition issued</span>
      </span>
      <div class="govuk-radios">
        <div class="govuk-radios__item" *ngFor="let option of prohibitionOptions">
          <input
            class="govuk-radios__input"
            id="test-prohibitionIssued-{{ option.id }}"
            formControlName="prohibitionIssued"
            type="radio"
            value="{{ option.selected }}"
            [checked]="option.selected"
            [class.govuk-input--error]="
              isSubmitted && testResultChildForm.form.get('prohibitionIssued')?.invalid
            "
          />
          <label class="govuk-label govuk-radios__label">
            {{ option.name }}
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
