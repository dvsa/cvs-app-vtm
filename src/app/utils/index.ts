export * from './mockStore';
export * from './test-model.utils';
export * from './testing.utils';
export * from './form.utils';


