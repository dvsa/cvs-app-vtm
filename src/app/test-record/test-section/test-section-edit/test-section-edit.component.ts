import { Component, OnInit, ChangeDetectionStrategy, Input } from '@angular/core';
import { TestType } from '@app/models/test.type';
import { TestResultModel } from '@app/models/test-result.model';
import { MatDialog } from '@angular/material/dialog';
import { DialogBoxConfirmationComponent } from '@app/shared/dialog-box-confirmation/dialog-box-confirmation.component';
import { Router } from '@angular/router';
import {
  REASON_FOR_ABANDONING_HGV_TRL,
  REASON_FOR_ABANDONING_PSV,
  REASON_FOR_ABANDONING_TIR,
  RESULT
} from '@app/test-record/test-record.enums';
import {
  ControlContainer,
  FormControl,
  FormGroup,
  FormGroupDirective,
  Validators
} from '@angular/forms';
import { Subscription } from 'rxjs';
import { DisplayOptionsPipe } from '@app/pipes/display-options.pipe';
import { SelectOption } from '@app/models/select-option';
import { TestRecordMapper } from '@app/test-record/test-record.mapper';

@Component({
  selector: 'vtm-test-section-edit',
  templateUrl: './test-section-edit.component.html',
  styleUrls: ['./test-section-edit.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  viewProviders: [{ provide: ControlContainer, useExisting: FormGroupDirective }]
})
export class TestSectionEditComponent implements OnInit {
  @Input() testType: TestType;
  @Input() testRecord: TestResultModel;
  @Input() applicableTestTypeIds1: {};
  @Input() applicableTestTypeIds2: {};
  @Input() isSubmitted: boolean;
  resultOptions: string[] = Object.values(RESULT);
  testResultChildForm: FormGroupDirective;
  testTypeGroup: FormGroup;
  testResultSubscription: Subscription;
  isAbandoned: boolean;
  prohibitionOptionSelected: string;
  prohibitionOptions: SelectOption[];
  reasonsForAbandoning: string[];

  constructor(
    private dialog: MatDialog,
    private router: Router,
    parentForm: FormGroupDirective,
    private testRecordMapper: TestRecordMapper
  ) {
    this.testResultChildForm = parentForm;
  }

  ngOnInit() {
    this.isAbandoned = this.testType.testResult === 'abandoned';
    this.reasonsForAbandoning = this.testRecordMapper.getReasonsForAbandoning(
      this.testRecord.vehicleType,
      this.testType.testTypeId
    );
    this.testTypeGroup = this.testResultChildForm.form.get('testType') as FormGroup;

    if (!!this.testTypeGroup) {

      this.testTypeGroup.addControl(
        'testResult',
        new FormControl(this.testType.testResult, Validators.required)
      );

      this.testTypeGroup.addControl(
        'reasonForAbandoning',
        new FormControl(this.testType.reasonForAbandoning, Validators.required)
      );

      this.testTypeGroup.addControl(
        'additionalCommentsForAbandon',
        new FormControl(this.testType.additionalCommentsForAbandon)
      );

      this.testTypeGroup.addControl(
        'certificateNumber',
        new FormControl(this.testType.certificateNumber, Validators.required)
      );

      this.testTypeGroup.addControl(
        'testExpiryDate',
        new FormControl(
          this.testType.testExpiryDate ? this.testType.testExpiryDate.split('T')[0] : '',
          Validators.required
        )
      );

      this.testTypeGroup.addControl(
        'testAnniversaryDate',
        new FormControl(
          this.testType.testAnniversaryDate
            ? this.testType.testAnniversaryDate.split('T')[0]
            : '',
          Validators.required
        )
      );

      this.testTypeGroup.addControl(
        'testEndTimestamp',
        new FormControl(
          this.testRecord.testEndTimestamp ? this.testRecord.testEndTimestamp.split('T')[0] : '',
          Validators.required
        )
      );

      this.testTypeGroup.addControl(
        'testTypeStartTimestamp',
        new FormControl('22:11', Validators.required)
      );

      this.testTypeGroup.addControl(
        'testTypeEndTimestamp',
        new FormControl('11:33', Validators.required)
      );

      this.testTypeGroup.addControl(
        'prohibitionIssued',
        new FormControl(this.testType.prohibitionIssued, Validators.required)
      );

      this.testResultSubscription = this.testTypeGroup
        .get('testResult')
        .valueChanges.subscribe((value) => {
          this.isAbandoned = value === 'abandoned';
        });
    }

    this.prohibitionOptionSelected = this.testType.prohibitionIssued ? 'Yes' : 'No';
    this.prohibitionOptions = new DisplayOptionsPipe().transform(
      ['Yes', 'No'],
      [this.prohibitionOptionSelected]
    );
  }

  onChangeTestType() {
    const dialogRef = this.dialog.open(DialogBoxConfirmationComponent, {
      width: '45vw',
      data: {
        context:
          '<h2 class="govuk-heading-l">Change test type</h2>' +
          '<p class="govuk-body">If you change the test type, some of the current test details will be lost.</br>' +
          'Print a version of the current test record or save the details to refer to when entering the details for the new test type.</p>',
        actionName: 'Change test type'
      }
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result) {
        this.router.navigate(['/select-test-type', this.testType.testNumber]);
      }
    });
  }
}
