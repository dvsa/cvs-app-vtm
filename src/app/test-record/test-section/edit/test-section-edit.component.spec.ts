import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { TestSectionEditComponent } from './test-section-edit.component';
import { SharedModule } from '@app/shared/shared.module';
import { TESTING_UTILS } from '@app/utils/testing.utils';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { RouterTestingModule } from '@angular/router/testing';
import { FormGroupDirective, ReactiveFormsModule } from '@angular/forms';
import {
  REASON_FOR_ABANDONING_HGV_TRL,
  REASON_FOR_ABANDONING_PSV
} from '@app/test-record/test-record.enums';
import { TEST_TYPE_APPLICABLE_UTILS } from '@app/utils/test-type-applicable-models.utils';
import { TEST_MODEL_UTILS } from '@app/utils';

describe('TestSectionEditComponent', () => {
  let component: TestSectionEditComponent;
  let fixture: ComponentFixture<TestSectionEditComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule, SharedModule, ReactiveFormsModule],
      declarations: [TestSectionEditComponent],
      providers: [
        {
          provide: FormGroupDirective,
          useValue: TESTING_UTILS.mockFormGroupDirective()
        }
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestSectionEditComponent);
    component = fixture.componentInstance;
    component.testRecord = TEST_MODEL_UTILS.mockTestRecord();
    component.testType = TEST_MODEL_UTILS.mockTestType({
      testResult: 'pass',
      testTypeStartTimestamp: '2020-01-16T12:24:38.027Z',
      reasonForAbandoning: 'test'
    });
    component.testTypesApplicable = TEST_TYPE_APPLICABLE_UTILS.mockTestTypesApplicable();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    fixture.detectChanges();
    expect(component.isAbandoned).toEqual(false);
    expect(component.reasonsForAbandoning).toEqual(Object.values(REASON_FOR_ABANDONING_HGV_TRL));
    expect(fixture).toMatchSnapshot();
  });

  it('should update isAbandoned based on test result value', () => {
    component.testType = TEST_MODEL_UTILS.mockTestType({
      testResult: 'abandoned',
      reasonForAbandoning: 'test'
    });
    fixture.detectChanges();
    expect(component.isAbandoned).toEqual(true);
  });

  it('should update the reasons for abandoning based on vehicle type', () => {
    component.testRecord = TEST_MODEL_UTILS.mockTestRecord({ vehicleType: 'psv' });
    fixture.detectChanges();
    expect(component.reasonsForAbandoning).toEqual(Object.values(REASON_FOR_ABANDONING_PSV));
  });
});
