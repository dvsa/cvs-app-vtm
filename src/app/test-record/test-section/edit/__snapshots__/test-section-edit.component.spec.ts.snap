// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestSectionEditComponent should create 1`] = `
<vtm-test-section-edit
  dialog={[Function MatDialog]}
  isAbandoned="false"
  prohibitionOptions={[Function Object]}
  reasonsForAbandoning={[Function Array]}
  reasonsForAbandoningOptions={[Function Array]}
  resultOptions={[Function Array]}
  router={[Function Router]}
  selectedOptions={[Function Array]}
  testRecord={[Function Object]}
  testRecordMapper={[Function TestRecordMapper]}
  testResult={[Function EventEmitter]}
  testResultChildForm={[Function FormGroupDirective]}
  testResultSubscription={[Function Subscriber]}
  testType={[Function Object]}
  testTypeGroup={[Function FormGroup]}
  testTypesApplicable={[Function Object]}
>
  <dl
    class="govuk-summary-list"
  >
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
         Test type 
      </dt>
      <dd
        class="govuk-summary-list__value"
        id="test-testTypeName"
      >
         First Test 
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <a
          class="govuk-link govuk-!-font-size-19"
          href="javascript:void(0)"
        >
          <span>
            Change
          </span>
        </a>
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
         Test code 
      </dt>
      <dd
        class="govuk-summary-list__value"
        id="test-testCode"
      >
         - 
      </dd>
      <dd
        class="govuk-summary-list__value"
      />
    </div>
  </dl><form
    class="ng-untouched ng-pristine ng-invalid"
    formgroupname="testType"
    ng-reflect-name="testType"
    novalidate=""
  >
    <div
      class="govuk-grid-row govuk-!-margin-bottom-3"
    >
      <div
        class="govuk-grid-column-full"
      >
        <label
          class="govuk-label govuk-!-font-weight-bold"
        >
           Result 
        </label>
        <div
          class="govuk-radios"
        >
          
          <div
            class="govuk-radios__item"
          >
            <input
              class="govuk-radios__input ng-untouched ng-pristine ng-valid"
              formcontrolname="testResult"
              id="test-testResult-Fail"
              ng-reflect-form-control-name="testResult"
              ng-reflect-name="testResult"
              ng-reflect-value="fail"
              type="radio"
            />
            <label
              class="govuk-label govuk-radios__label"
            >
               Fail 
            </label>
          </div>
          <div
            class="govuk-radios__item"
          >
            <input
              class="govuk-radios__input ng-untouched ng-pristine ng-valid"
              formcontrolname="testResult"
              id="test-testResult-Pass"
              ng-reflect-form-control-name="testResult"
              ng-reflect-name="testResult"
              ng-reflect-value="pass"
              type="radio"
            />
            <label
              class="govuk-label govuk-radios__label"
            >
               Pass 
            </label>
          </div>
          <div
            class="govuk-radios__item"
          >
            <input
              class="govuk-radios__input ng-untouched ng-pristine ng-valid"
              formcontrolname="testResult"
              id="test-testResult-PRS"
              ng-reflect-form-control-name="testResult"
              ng-reflect-name="testResult"
              ng-reflect-value="prs"
              type="radio"
            />
            <label
              class="govuk-label govuk-radios__label"
            >
               PRS 
            </label>
          </div>
          <div
            class="govuk-radios__item"
          >
            <input
              class="govuk-radios__input ng-untouched ng-pristine ng-valid"
              formcontrolname="testResult"
              id="test-testResult-Abandoned"
              ng-reflect-form-control-name="testResult"
              ng-reflect-name="testResult"
              ng-reflect-value="abandoned"
              type="radio"
            />
            <label
              class="govuk-label govuk-radios__label"
            >
               Abandoned 
            </label>
          </div>
        </div>
      </div>
    </div>
    
    
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
           Test number 
        </dt>
        <dd
          class="govuk-summary-list__value"
          id="test-testNumber"
        >
           123123 
        </dd>
        <dd
          class="govuk-summary-list__value"
        />
      </div>
    </dl>
    
    
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
           Test date 
        </dt>
        <dd
          class="govuk-summary-list__value"
          id="test-testTypeEndDate"
        >
          
          <span>
             16/01/2020 
          </span>
        </dd>
        <dd
          class="govuk-summary-list__value"
        />
      </div>
    </dl>
    <div
      class="govuk-grid-row govuk-!-margin-bottom-3"
    >
      <div
        class="govuk-grid-column-three-quarters"
      >
        <label
          class="govuk-label govuk-!-font-weight-bold"
        >
           Start time 
        </label>
        <vtm-time-input
          ariadescribedby="testTypeStartTimestamp"
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="testTypeStartTimestamp"
          id="test-testTypeStartTime"
          ng-reflect-name="testTypeStartTimestamp"
          ngdefaultcontrol=""
        />
      </div>
    </div>
    <div
      class="govuk-grid-row govuk-!-margin-bottom-3"
    >
      <div
        class="govuk-grid-column-three-quarters"
      >
        <label
          class="govuk-label govuk-!-font-weight-bold"
        >
           End time 
        </label>
        <vtm-time-input
          ariadescribedby="testTypeEndTimestampTime"
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="testTypeEndTimestampTime"
          id="test-testTypeEndTime"
          ng-reflect-name="testTypeEndTimestampTime"
          ngdefaultcontrol=""
        />
      </div>
    </div>
    
  </form>
</vtm-test-section-edit>
`;
